src/modules/src/controller_indi.c

```diff
@@ -0,0 +1,311 @@
+/*
+ *
+ * Copyright (c) 2019 Ewoud Smeur and Andre Luis Ogando Paraense
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, in version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * This control algorithm is the Incremental Nonlinear Dynamic Inversion (INDI)
+ * controller.
+ *
+ * This is an implementation of the publication in the
+ * journal of Control Guidance and Dynamics: Adaptive Incremental Nonlinear
+ * Dynamic Inversion for Attitude Control of Micro Aerial Vehicles
+ * http://arc.aiaa.org/doi/pdf/10.2514/1.G001490
+ */
+
+#include "controller_indi.h"
+
+struct IndiVariables indi = {
+  .max_rate = STABILIZATION_INDI_MAX_RATE,
+  .attitude_max_yaw_rate = STABILIZATION_INDI_MAX_R,
+
+  .g1 = {STABILIZATION_INDI_G1_P, STABILIZATION_INDI_G1_Q, STABILIZATION_INDI_G1_R},
+  .g2 = STABILIZATION_INDI_G2_R,
+  .reference_acceleration = {
+    STABILIZATION_INDI_REF_ERR_P,
+    STABILIZATION_INDI_REF_ERR_Q,
+    STABILIZATION_INDI_REF_ERR_R,
+    STABILIZATION_INDI_REF_RATE_P,
+    STABILIZATION_INDI_REF_RATE_Q,
+    STABILIZATION_INDI_REF_RATE_R
+  },
+
+  /* Estimation parameters for adaptive INDI */
+  .est = {
+    .g1 = {
+      STABILIZATION_INDI_G1_P / INDI_EST_SCALE,
+      STABILIZATION_INDI_G1_Q / INDI_EST_SCALE,
+      STABILIZATION_INDI_G1_R / INDI_EST_SCALE
+    },
+    .g2 = STABILIZATION_INDI_G2_R / INDI_EST_SCALE,
+    .mu = STABILIZATION_INDI_ADAPTIVE_MU,
+  },
+  .adaptive = STABILIZATION_INDI_USE_ADAPTIVE,
+};
+
+static inline void float_rates_zero(struct FloatRates *fr) {
+	fr->p = 0.0f;
+	fr->q = 0.0f;
+	fr->r = 0.0f;
+}
+
+void indi_init_filters(void)
+{
+  // tau = 1/(2*pi*Fc)
+  float tau = 1.0f / (2.0f * PI * STABILIZATION_INDI_FILT_CUTOFF);
+  float tau_r = 1.0f / (2.0f * PI * STABILIZATION_INDI_FILT_CUTOFF_R);
+  float tau_axis[3] = {tau, tau, tau_r};
+  float tau_est = 1.0f / (2.0f * PI * STABILIZATION_INDI_ESTIMATION_FILT_CUTOFF);
+  float sample_time = 1.0f / PERIODIC_FREQUENCY;
```

Comment:
You can see that for this discrete filter, the control frequency is taken into account. Therefore, the cutoff can be specified independent of the control frequency. For the actuator dynamics I should also do that.

Reply:
Did not understand if action is needed here or not.

Reply:
Nope 😉